graphix/pattern.py

```diff
@@ -421,10 +472,39 @@ def shift_signals(self, method="local") -> dict[int, list[int]]:
                 else:
                     self._commute_with_following(target)
                 target += 1
-        else:
-            raise ValueError("Invalid method")
         return swapped_dict
 
+    def shift_signals_direct(self) -> dict[int, set[int]]:
+        signal_dict = {}
+
+        def expand_domain(domain: set[command.Node]) -> None:
+            for node in domain & signal_dict.keys():
+                domain ^= signal_dict[node]
+
+        for cmd in self:
+            if cmd.kind == CommandKind.M:
+                s_domain = cmd.s_domain
+                t_domain = cmd.t_domain
+                expand_domain(s_domain)
+                expand_domain(t_domain)
+                plane = cmd.plane
+                if plane == Plane.XY:
+                    if t_domain:
+                        cmd.t_domain = set()
+                        signal_dict[cmd.node] = t_domain
+                elif plane == Plane.XZ:
+                    if s_domain:
+                        cmd.s_domain = set()
+                        t_domain ^= s_domain
+                        signal_dict[cmd.node] = s_domain
+                elif plane == Plane.YZ:
+                    if s_domain:
+                        cmd.s_domain = set()
+                        signal_dict[cmd.node] = s_domain
+            elif cmd.kind == CommandKind.X or cmd.kind == CommandKind.Z:
```

Comment:
@thierry-martinez 
Could you show me a reference or proof of this part?
I tried to use this method for a pattern containing XZ plane, but the result seems not deterministic.

Reply:
I added proofs in comments in e539c42.

Reply:
I got what is the cause of bug on my side. You are using different definitions of s_domain and t_domain right? In the original paper(Danos 2007), they defined s_domain as sign change of angle and t_domain as a addition of pi(1). Why do we follow this non-standard way? Or is there any reference defining s_domain and t_domain as they are later?

Anyway, I have ensured that the algorithm works well in the current convention. I approve PR!

Reply:
If I understand correctly, Danos 2007 only handles measurement on the XY plane, where the two conventions coincide: $M^{XY,α} X^s Z^t = M^{XY,(-1)^s·α+tπ}$. In the graphix implementation, we choose to follow the left-hand side of this equation for defining the effect of $s$-domain and $t$-domain on other planes, by defining $_t[M^{\lambda,\alpha}]^s = M^{\lambda,α} X^s Z^t$.

Reply:
I understand what you mean, and the LHS definition seems natural for the measurement calculus. However, didn't they mathematically choose the RHS definition as measurement command in the original context?

If they updated their defition in the extended measurement calculus paper, I'm willing to accept the current convention. Unfortunately, I don't have access to that. @shinich1, could you please check it?

Reply:
@masa10-f unfortunately extended MC paper '09 discuss YZ and XZ measurement commands in passing, without any mention of signals associated and the commutaion relations in the presence of signals. Backens '21 seem to introduce in RHS format for all plane.

Reply:
Indeed, in Backens 2021, p. 8, measure commands are defined as $M^{\lambda, \gamma(\alpha, s, t)}$, where $\gamma(\alpha, s, t) = (-1)^s \cdot \alpha + t \pi$. Thanks for pointing out the reference! Should we open another issue to discuss these conventions? This PR doesn't change the existing conventions and maintains the behavior already implemented in graphix, just more efficiently and concisely. Therefore, I believe it can be merged, even if our conventions may be subject to change later.

Reply:
@shinich1 Thanks!

I agree to discuss it in another issue, as it is not the point of this PR